flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/StreamTaskTest.java

```diff
@@ -1528,6 +1529,66 @@ public void testProcessWithUnAvailableInput() throws Exception {
         }
     }
 
+    @Test
+    public void testRestorePerformedOnlyOnce() throws Exception {
+        // given: the operator with empty snapshot result (all state handles are null)
+        OneInputStreamOperator<String, String> statelessOperator =
+                streamOperatorWithSnapshot(new OperatorSnapshotFutures());
+
+        try (MockEnvironment mockEnvironment = new MockEnvironmentBuilder().build()) {
+
+            // when: Invoke the restore explicitly before launching the task.
+            RunningTask<StreamTask<?, ?>> task =
+                    runTask(
+                            () -> {
+                                StreamTask mockStreamTask =
+                                        spy(
+                                                new MockStreamTaskBuilder(mockEnvironment)
+                                                        .setStreamInputProcessor(
+                                                                new AvailabilityTestInputProcessor(
+                                                                        10))
+                                                        .build());
+
+                                mockStreamTask.restore();
+
+                                return mockStreamTask;
+                            });
+            waitMailBoxIsRunning(task.streamTask);
+
+            task.streamTask.cancel();
+
+            // then: 'restore' was called only once.
+            verify(task.streamTask).restore();
```

Comment:
We shouldn't be using mockito. Section "Avoid Mockito - Use reusable test implementations" from:
https://flink.apache.org/contributing/code-style-and-quality-common.html#design-for-testability